addons/stock/models/product.py

```diff
@@ -759,6 +759,18 @@ def _onchange_type(self):
         return res
 
     def write(self, vals):
+        if 'company_id' in vals and vals['company_id']:
+            for product in self:
+                if product.company_id.id == vals['company_id']:
+                    continue
+            # Forbids to change the product's company it it exists quants in other company.
+            quants = self.env['stock.quant'].search([
+                ('product_id', 'in', product.product_variant_ids.ids),
+                ('company_id', '!=', vals['company_id']),
+            ])
+            if quants.exists():
```

Comment:
```suggestion
            if quants:
```
Search returns already only existing records